### PR TITLE
Pluggable runners

### PR DESCRIPTION
**Pull request by bigmstone**
This will add the ability to treat runners as plugins which should simplify the development process for creating new runners where needed.

- [x] Move runners to and load runners from contrib
- [x] Move existing dictionary to individual manifest files describing each runner
- [x] Modify action execution to call runners from new location
- [x] Finish Mistral Query
- [x] Acceptance testing
- [x] Add testing where needed
- [x] Get tests passing
- [x] Clean up code/add comments where needed/prep for master

### CODE REVIEW

File: contrib/runners/actionchainrunner/runner.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+- aliases: []
+  description: A runner for launching linear action chains.
+  enabled: true
+  name: action-chain
+  runner_module: st2actions.runners.actionchainrunner
+  runner_parameters:
```
Comment (lakshmi-kannan) [inline contrib/runners/actionchainrunner/runner.yaml]: We should call it parameters and in the code we can call it runner_parameters. UX feedback.

File: contrib/runners/actionchainrunner/runner.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+- aliases: []
+  description: A runner for launching linear action chains.
+  enabled: true
+  name: action-chain
+  runner_module: st2actions.runners.actionchainrunner
```
Comment (lakshmi-kannan) [inline contrib/runners/actionchainrunner/runner.yaml]: We call it entry_point for other content. But if you are going to use stevedore approach, some of this would change.

File: contrib/runners/cloudslang/runner.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+- aliases: []
+  description: A runner to execute cloudslang flows.
+  enabled: true
+  name: cloudslang
```
Comment (m4dcoder) [inline contrib/runners/cloudslang/runner.yaml]: Can we drop cloud slang and move it to st2contrib?

**Comment by m4dcoder** [comment]
We should also abstract callback handlers for the runners (i.e. mistral callback handler).

File: st2common/st2common/content/loader.py [hunk 4]
```diff
@@ -204,6 +240,36 @@ def _get_folder(self, pack_dir, content_type):
             return None
         return path
 
+    def _get_runners_from_dir(self, base_dir):
+        result = {}
+        for runner_name in os.listdir(base_dir):
```
Comment (Kami) [inline st2common/st2common/content/loader.py]: Perhaps we should sort the `os.listdir` result to always register runners in the same order? Or maybe it's not a big deal.

File: st2common/st2common/content/loader.py [hunk 5]
```diff
@@ -204,6 +240,36 @@ def _get_folder(self, pack_dir, content_type):
...
+            runner_dir = os.path.join(base_dir, runner_name)
+            runner_manifest_file = os.path.join(runner_dir, RUNNER_MANIFEST_FILE_NAME)
+
+            if os.path.isdir(runner_dir) and os.path.isfile(runner_manifest_file):
```
Comment (Kami) [inline st2common/st2common/content/loader.py]: Probably a good idea to add a DEBUG log statement when we found a valid directory and also when we don't (e.g. missing manifest file).

File: st2common/st2common/content/loader.py (hunk 4 above)
Comment (Kami) [inline st2common/st2common/content/loader.py]: Do we need to do some sanitation with the directory name (we do it for packs)?

Comment (Kami) [inline st2common/st2common/content/loader.py]: On a related note, we have a similar "issue" with packs - we actually use pack directory name as pack name ID, even though there is a "name" attribute in the metadata file (the latter is only used for display in API).

It's slightly confusing, but yeah for consistency we should probably do the same for runners, but the problem is runners names right now contain dashes, but for consistency with packs, etc. we should only allow underscores in the directory name.

Anyway, we can probably come up with some compromise, similar as we do with pack name alias right now for backward compatibility.

File: contrib/runners/remote_command_runner/runner.yaml [hunk 6]
```diff
@@ -0,0 +1,80 @@
+- aliases:
+  - run-remote
```
Comment (Kami) [inline contrib/runners/remote_command_runner/runner.yaml]: While we are at it, we should also get rid of the old aliases which have been deprecated for a while now.

Let's just make sure upgrade notes is updated and all the code in st2contrib as well (if there is any action which still uses old and deprecated runner name).

File: contrib/runners/mistral/runner.yaml [hunk 7]
```diff
@@ -2,8 +2,8 @@
   description: A runner for executing mistral v2 workflow.
   enabled: true
   name: mistral-v2
-  query_module: st2actions.query.mistral.v2
-  runner_module: st2actions.runners.mistral.v2
+  query_module: mistralv2
+  runner_module: mistralv2
```
Comment (Kami) [inline contrib/runners/mistral/runner.yaml]: Does the new infrastructure support sub-directories?

E.g. mistral.v2 would map to `<runners path>/mistral/v2.py` or similar.

If not, it would be nice if we can also make support that, but in the worst case we can call it mistral_v2 for now :)

File: contrib/runners/pythonrunner/pythonrunner.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
```
Comment (Kami) [inline contrib/runners/pythonrunner/pythonrunner.py]: Another thing - let's make sure that those files are still linted (might need to update lint targets in the makefile, but we already lint some other Python files in that directory, might just need to add contrib/runners/*/.py to it or similar).

**Comment by bigmstone** [comment]
After fighting with my env most of today to get tests running faster I am stuck at how to best move the tests around for runners.

One problem is the runners are in `contrib` now so they must be imported by explicit path. I could use the loader to do this, but the loader by default points to the base_dir known by st2_config. So I'd have to have a loader for tests that uses relative paths. Not the end of the world, but not ideal.

Second problem is using mock's patch decorator. It takes module import path as a string. Don't have a solution for this one yet, so open to suggestions.

I also lean towards having runner test in the directory for the runner. But, assuming that's not part of the solution to the above, doesn't have to be done in this PR.

**Comment by bigmstone** [comment]
I've worked through most of the testing issues. Still have a few hanging around. Any insights welcome.

**Comment by bigmstone** [comment]
The tests are expecting the action to be in a different state from `succeeded` but they're all succeeding (Instead of `failed`, `delayed`, etc.). Only tests that are failing have the `mock.patch.object` of the `LocalShellRunner` object in common. I've looked over all the changes I've made now 3 times and can't find anything that would screw up concurrency, but that's where my gut is taking me right now. Will continue looking, but would love an extra set of eyes.

**Comment by m4dcoder** [comment]
@bigmstone here's a pointer to the problem you're seeing.  the mock @ https://github.com/StackStorm/st2/blob/master/st2actions/tests/unit/policies/test_concurrency.py#L59 is suppose to keep the action in a running state.  this makes sure we can test the threshold and concurrency.  in this branch, the action executions are moving to succeeded state.   so the changes has affected how that mock is suppose to work.  i'll let you pick this up from here.

**Comment by bigmstone** [comment]
After much digging it would appear I'm no longer patching the same instance of `LocalSSHRunner` which is a side effect of loading the runners via `imp`. So I need a way to patch the instance of the runner being imported by the runner loader.

**Comment by bigmstone** [comment]
Tests finally nailed down. (Halalurrya)

Still TODO:
* Finish Mistral Query
* Acceptance testing
* Clean up/Add tests

Maybe, needs discussion on if we do this in the current PR:
* Migrate away from `register_runner` and use existing loader to return class.

File: st2actions/st2actions/runners/pythonrunner.py [hunk 9]
```diff
@@ -1,272 +0,0 @@
-# Licensed to the StackStorm, Inc ('StackStorm') under one or more
```
Comment (Kami) [inline st2actions/st2actions/runners/pythonrunner.py]: Yeah we need to verify the symlink works as expected end to end (it might cause issues with virtual environments...).

One option would also be to have a Python module here which basically just imports everything from new location and re-exports it, so:

```python
__all__ = [
    'Action'
]


from st2actions.runners.base_action import Action
```

This is probably also better from cross-OS compatibility perspective :)

Comment (bigmstone) [inline st2actions/st2actions/runners/pythonrunner.py]: Yes, I like this. I also think I will do this for the test runners. This is much cleaner than throwing symlinks around everywhere.

File: st2common/st2common/content/loader.py [hunk 10]
```diff
@@ -15,22 +15,59 @@
...
 __all__ = [
+    'ContentRunnerLoader',
     'ContentPackLoader',
     'MetaLoader'
 ]
 
 LOG = logging.getLogger(__name__)
 
 
+class ContentRunnerLoader(object):
```
Comment (bigmstone) [inline st2common/st2common/content/loader.py]: As discussed stevedore is -1 and at the very least not happening in this PR. (I realize this is already decided, but I'm leaving comments on all existing comments so we can get this ready to close out.)

File: contrib/runners/cloudslang/runner.yaml (hunk 3 above)
Comment (bigmstone) [inline contrib/runners/cloudslang/runner.yaml]: IMO this PR is big enough. I agree this should/could be dropped, but should be done in a different PR.

File: st2common/st2common/content/loader.py (hunk 4 above)
Comment (bigmstone) [inline st2common/st2common/content/loader.py]: I can't conceive of a time when this would be a big deal, but would be trivial to implement.

File: st2actions/st2actions/runners/pythonrunner.py (hunk 9 above)
Comment (bigmstone) [inline st2actions/st2actions/runners/pythonrunner.py]: Done, and unfortunately can't do it for test fixtures because of manifest file.

File: st2common/st2common/content/loader.py (hunk 5 above)
Comment (bigmstone) [inline st2common/st2common/content/loader.py]: Done

File: contrib/runners/remote_command_runner/runner.yaml (hunk 6 above)
Comment (bigmstone) [inline contrib/runners/remote_command_runner/runner.yaml]: Done

Comment (bigmstone) [inline contrib/runners/remote_command_runner/runner.yaml]: I removed this, but it had a cascading effect. Would result in a few more files being changed. I changed it back because I think this PR is big enough as it is. Can make this change in another PR.

File: contrib/runners/mistral/runner.yaml (hunk 7 above)
Comment (bigmstone) [inline contrib/runners/mistral/runner.yaml]: It doesn't currently support sub-directories because the modules are loaded by `<base_path>/runners/<runner_module>/<runner_module>.py`. We could change it to use `<base_path>/runners/<runner_name>/<runner_module>.py`

What is the goal? Is the goal to load various modules from the same folder or to be able to have sub-directories inside of runners? For instance `<base_path>/runners/<runner_name>/<some>/<new>/<path>/<runner_module>.py`?

File: st2common/st2common/content/loader.py (hunk 4 above)
Comment (bigmstone) [inline st2common/st2common/content/loader.py]: We have runner alias as well. So it would be trivial to name, for instance, `mistral-v2` to `mistral_v2` and add an alias for `mistral-v2`. This also ties into [your other comment](https://github.com/StackStorm/st2/pull/2824#pullrequestreview-967994).

What type sanitation of the directory name are you thinking? Enforcing not using dash, for instance? I didn't see anything obvious in `loader.py` that adds any additional sanitation for pack loading.

File: st2common/st2common/content/loader.py (hunk 10 above)
Comment (Kami) [inline st2common/st2common/content/loader.py]: Maybe just RunnersLoader / RunnerMetadataLoader or similar is a better name? "ContentRunner" makes it sound a bit confusing, at least to me.

With prefix pack loader with content because we used to call them "content packs" (these days days we mostly just call them packs or integration packs).

File: st2common/st2common/constants/runners.py [hunk 11]
```diff
@@ -44,7 +49,7 @@
 
 REMOTE_RUNNER_PRIVATE_KEY_HEADER = 'PRIVATE KEY-----'.lower()
 
-# Python runner
+# Python runner'
```
Comment (Kami) [inline st2common/st2common/constants/runners.py]: How did this ' get here :P

File: st2common/st2common/content/loader.py (hunk 4 above)
Comment (Kami) [inline st2common/st2common/content/loader.py]: Yep, not allowing dash at least.

Here is a regex we use for pack names - `PACK_NAME_WHITELIST = r'^[A-Za-z0-9_-]+'`.

File: st2common/st2common/config.py [hunk 12]
```diff
@@ -75,11 +75,16 @@ def register_opts(ignore_errors=False):
...
+    system_runners_base_path = os.path.join(cfg.CONF.system.base_path, 'runners')
     content_opts = [
         cfg.StrOpt('system_packs_base_path', default=system_packs_base_path,
                    help='Path to the directory which contains system packs.'),
+        cfg.StrOpt('system_runners_base_path', default=system_runners_base_path,
+                   help='Path to the directory which contains system runners.'),
         cfg.StrOpt('packs_base_paths', default=None,
-                   help='Paths which will be searched for integration packs.')
+                   help='Paths which will be searched for integration packs.'),
+        cfg.StrOpt('runners_base_paths', default=None,
```
Comment (Kami) [inline st2common/st2common/config.py]: I'm fine with two settings for consistency with packs, but I'm also OK with just having a single setting for simplicity (user runners can just live in the same directory as system ones).

Anyways, it's fine as it is now.

File: st2common/st2common/content/bootstrap.py [hunk 13]
```diff
@@ -57,6 +57,7 @@ def register_opts():
         cfg.BoolOpt('configs', default=False, help='Register and load pack configs.'),
 
         cfg.StrOpt('pack', default=None, help='Directory to the pack to register content from.'),
+        cfg.StrOpt('runner', default=None, help='Directory to load runners from.'),
```
Comment (Kami) [inline st2common/st2common/content/bootstrap.py]: Hm, should this be called `runners-path` or similar instead?

File: st2common/st2common/runners/__init__.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
```
Comment (Kami) [inline st2common/st2common/runners/__init__.py]: Minor thing, I would prefer to put this into `base.py` file and leave __init__.py empty.

Comment (bigmstone) [inline st2common/st2common/runners/__init__.py]: Would you like this changed in this PR or a later one? I was just leaving it as it was for now, but happy to change it.

File: st2common/st2common/content/bootstrap.py (hunk 13 above)
Comment (bigmstone) [inline st2common/st2common/content/bootstrap.py]: That makes more sense to me, I was just following the contention set by `pack`.

File: st2common/st2common/content/loader.py (hunk 10 above)
Comment (bigmstone) [inline st2common/st2common/content/loader.py]: Done. Will push in next commit. (Putting this here now to keep track)

File: st2common/st2common/content/loader.py (hunk 4 above)
Comment (bigmstone) [inline st2common/st2common/content/loader.py]: Got it. Didn't break open the `setup_pack_virtualenv` function which would appear where we do this (In addition to when you setup venv with pack mgmt).

**Comment by Kami** [comment]
Thanks.

It looks good to me right now, but before merging I would prefer to spend some time on it and test it locally. In addition to that, I think we should merge it once at least one of us has a day to spend on the things which might break after the merge (aka end to end stuff).

**Comment by bigmstone** [comment]
@kami agree. I won't get a chance to work on it this weekend but I will plan to start this on Monday. I'm going to begin work on the new runner next week so I can have both going at the same time. If you or someone else can do the same I would appreciate it. If we find anything I can spend time fixing it and will plan to be ready to focus exclusively on it if needed when we merge to master.

File: contrib/runners/remote_command_runner/runner.yaml [hunk 15]
```diff
@@ -15,11 +15,11 @@
       description: Arbitrary Linux command to be executed on the remote host(s).
       type: string
     cwd:
-      default: "~/"
+      default: "/tmp"
```
Comment (Kami) [inline contrib/runners/remote_command_runner/runner.yaml]: How does this change affect things?

I would imagine it's backward incompatible in some scenarios - /tmp is not necessarily always accessible to all the users. In any case, I believe it should be at least fine for our bootstrap installations.

File: st2common/st2common/content/bootstrap.py [hunk 16]
```diff
@@ -172,19 +194,10 @@ def register_actions():
...
         LOG.info('############## Registering actions ######################')
         LOG.info('=========================================================')
-        runners_registrar.register_runner_types(experimental=cfg.CONF.experimental)
-    except Exception as e:
-        LOG.warning('Failed to register runner types: %s', e, exc_info=True)
-        LOG.warning('Not registering stock runners .')
-        return
-
-    # 2. Register actions
-    try:
```
Comment (bigmstone) [inline st2common/st2common/content/bootstrap.py]: https://github.com/StackStorm/st2/commit/0c5f5aab8fc8ab819889a65d9dab6bca815c45d7#diff-499d9f5cfe40e44d821453ced745e3b1R352
